tools/summaries/palm.py

```diff
@@ -12,49 +12,112 @@
+Do not include facts from other sources.
+Do not use superlatives.
+Do not use the phrase 'According to the data'.
+Do not include opinions.
+Please include references if information is included from other sources.
+Please write in a professional and business-neutral tone.
+
+{place_type}: {place_name}
+
+{ranking_key}:
+{ranking_data}
+
+Table:
+{data_table}
+
+Summary:
+"""
+
+_RESUMMARIZE_PROMPT = """
+Summarize these facts into 1 paragraph.
+Start by introducing the place.
+The summary should only be based on the information presented in these facts.
+Please write in a professional and business-neutral tone.
+
+Facts:
+{facts}
+
+Summary:
+"""
 
 assert _API_KEY, "$PALM_API_KEY must be specified."
 
+# Ranking key -> data_table key
+_TABLE_KEYS = {
+  "Largest Population": "Count_Person",
+  "Highest Median Income": "Median_Income_Person",
+  "Highest Median Age": "Median_Age_Person",
+}
+
 
-def get_summary(name: str, csv: str):
+def strip_superlatives(label: str) -> str:
+  label = label.replace("Highest", "")
+  label = label.replace("Largest", "")
+  return label.strip()
+
+
+def request_palm(prompt):
+  """Sends a summary request to PALM"""
   url = f"{_API_URL}?key={_API_KEY}"
-  prompt = f"Summarize the following CSV for {name} in 1 sentence. Don't use superlatives.\n\n{csv}"
-  # prompt = f"Give me a one sentence NL summary based on the following info for {name}:\n\n{csv}"
-  params = {
-      "prompt": {
-          "text": prompt
-      },
-      "temperature": _TEMPERATURE,
-      "candidateCount": 1,
-  }
   headers = {
-      "Content-Type": "application/json",
+    "Content-Type": "application/json",
+  }
+  params = {
+    "prompt": {
+      "text": prompt,
+    },
+    "temperature": _TEMPERATURE,
+    "candidateCount": 1,
   }
+  response = requests.post(url=url, json=params, headers=headers).json()
+  return response.get("candidates", [{}])[0].get("output", "")
+
+
+def get_summary(place_name: str, place_type: str, rankings: str, data_tables: List[str]):
+  """Generates an overview summary for a place"""
+  candidates = []
+  prompts = []
+  for ranking_key, data_table_key in _TABLE_KEYS.items():
+    if not ranking_key in rankings:
+      logging.info(f"Skipping {ranking_key} for {place_name}")
+      continue
+    if not data_table_key in data_tables:
+      logging.info(f"Skipping {data_table_key} for {place_name}")
+      continue
+    prompt_keys = {
+      "place_type": place_type,
+      "place_name": place_name,
+      "ranking_key": strip_superlatives(ranking_key),
+      "ranking_data": '\n'.join(rankings[ranking_key]),
+      "data_table": data_tables[data_table_key]
+    }
+    prompt = _SERIES_PROMPT.format(**prompt_keys)
+    prompts.append(prompt)
+
+    response = request_palm(prompt)
+    candidates.append("- " + response)
+
+    # TODO: Add some verification step here
+
+  facts = '\n'.join(candidates)
+  prompt = _RESUMMARIZE_PROMPT.format(facts=facts)
+  response = request_palm(prompt)
```

Comment:
This is very interesting. So we first summarize each table separately and then use those to resummarize. Very creative!

One thing I worry about is that it can add opinion on multiple occasions. What have you seen on that front?

Reply:
I've mostly observed opinions added to the first set of summaries. The resummarization is usually fairly straightforward and makes it look less templatized.